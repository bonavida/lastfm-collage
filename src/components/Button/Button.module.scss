@import 'styles/utils.scss';

.button {
  display: inline-block;
  font-weight: $font-weight-regular;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  user-select: none;
  color: $color-text-white-base;
  background-color: $color-primary;
  border: 1px solid transparent;
  line-height: $line-height-m;
  border-radius: $border-radius-rounded;
  font-size: $font-size-s;
  letter-spacing: $letter-spacing-l;
  text-transform: uppercase;
  will-change: transform;
  padding: $padding-sm $padding-xl;
  transform: perspective(1px) scale(1) translateZ(0);
  transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
    border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out,
    transform 0.15s ease-in;

  &:hover {
    background-color: $color-primary-dark;
    border-color: $border-color-primary-dark;
    transform: perspective(1px) scale(1.01) translateZ(0);
  }

  &:not(:disabled) {
    cursor: pointer;
  }

  &:focus {
    outline: 0;
    text-decoration: none;
    box-shadow: $box-shadow-primary-button;
  }

  &Secondary {
    background-color: #333;
    color: white;

    &:hover {
      background-color: #444;
      border-color: #000;
    }

    &:focus {
      box-shadow: $box-shadow-secondary-button;
    }
  }
}
