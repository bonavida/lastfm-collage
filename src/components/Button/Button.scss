@import 'styles/utils.scss';

.button {
  display: inline-block;
  font-size: $font-size-m;
  font-weight: $font-weight-regular;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  user-select: none;
  color: $color-text-white-base;
  background-color: $color-primary;
  border: 1px solid transparent;
  padding: $padding-s $padding-l;
  line-height: $line-height-m;
  border-radius: $border-radius-regular;
  transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
    border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  
  &:hover {
    background-color: $color-primary-dark;
    border-color: $border-color-primary-dark;
  }

  &:not(:disabled) {
    cursor: pointer;
  }

  &:focus {
    outline: 0;
    text-decoration: none;
    box-shadow: $box-shadow-primary-button;
  }
}
