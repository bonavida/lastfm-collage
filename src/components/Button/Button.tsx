import React from 'react';
/** Styles */
import './Button.scss';

type ButtonProps = {
  children: React.ReactNode;
  className?: string;
  disabled?: boolean;
  onClick: () => void;
};

const Button = ({
  children,
  className,
  disabled = false,
  onClick,
}: ButtonProps) => {
  return (
    <button
      className={`button ${className || ''}`}
      type="button"
      disabled={disabled}
      onClick={onClick}
    >
      {children}
    </button>
  );
};

export default Button;
