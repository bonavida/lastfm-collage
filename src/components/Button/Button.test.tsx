import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { axe } from 'jest-axe';
/** Components */
import Button from './Button';

type ButtonProps = {
  children: React.ReactNode;
  className?: string;
  props?: any;
  onClick: any;
};

const onClickMock = jest.fn();

const renderButton = (props: Partial<ButtonProps> = {}) => {
  const defaultProps: ButtonProps = {
    children: <span>Button test</span>,
    className: 'externalClassName',
    onClick: onClickMock,
  };
  return render(<Button {...defaultProps} {...props} />);
};

describe('<Button>', () => {
  afterEach(() => {
    cleanup();
  });

  test('Snapshot renders => it should be rendered with the provided props', () => {
    const { container } = renderButton();
    expect(container).toMatchSnapshot();
  });

  test('should fire onClick event on click over the button', () => {
    const { getByRole } = renderButton();

    fireEvent.click(getByRole('button'));

    expect(onClickMock).toHaveBeenCalled();
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  test('should not have basic accessibility issues', async () => {
    const { container } = renderButton();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
