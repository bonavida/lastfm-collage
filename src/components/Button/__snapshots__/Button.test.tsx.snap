// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> Snapshot renders => it should be rendered with the provided props 1`] = `
<div>
  <button
    class="button externalClassName"
    type="button"
  >
    <span>
      Button test
    </span>
  </button>
</div>
`;
